Sources/PaymentsCore/Networking/APIClient.swift

```diff
@@ -4,22 +4,42 @@ import Foundation
 ///
 /// `APIClient` is the entry point for each payment method feature to perform API requests. It also offers convenience methods for API requests used across multiple payment methods / modules.
 public class APIClient {
+        
+    // MARK: - Internal Properties
 
     private var http: HTTP
     private let sessionID = UUID().uuidString.replacingOccurrences(of: "-", with: "")
     private let coreConfig: CoreConfig
+    
+    /// The `AnalyticsService` instance is static/shared so that only one sessionID is used.
+    /// The "singleton" has to be managed here because `AnalyticsService` has a dependency on `HTTP`.
+    ///
+    /// Exposed for testing.
+    weak var analyticsService: AnalyticsService? {
+        get { APIClient._analyticsService }
+        set { APIClient._analyticsService = newValue }
+    }
+    private static var _analyticsService: AnalyticsService?
```

Comment:
Why cant we have a method like `AnalyticsClient.getInstance(http)`? If an instance has already been created, the `getInstance` method sets the new `http` to the instance. 
It feels kinda weird to manage it here. If we need it in another class, we have to repeat this code

Reply:
@scannillo is this the standard of doing singleton in swift?
since sessionID is now a one time set in `APIClient` we don't need to make `APIClient` to be singleton

Reply:
> Why cant we have a method like AnalyticsClient.getInstance(http)?

This is what I was hoping to do, initially too. But, this can't be done in Swift using the [typical singleton implementation](https://cocoacasts.com/what-is-a-singleton-and-how-to-create-one-in-swift](https://www.hackingwithswift.com/example-code/language/what-is-a-singleton)), because of the requirement for a private init. Therefore, you can't have a singleton with dependency injection.

> It feels kinda weird to manage it here. If we need it in another class, we have to repeat this code.

Yeh, I agree it's weird to have to manage the singleton this way. I know Jax and Matt did a lot of work to think of a cleaner way of doing this on BT too ([see thread](https://github.com/braintree/braintree_ios/pull/887#discussion_r1010550537)), but landed on this being the most straightforward. I'll keep messing around with ideas for other solutions.

Though, we shouldn't need to use this in any other place. The intention / design is to only have a single point of entry, via `APIClient.sendEvent`.

Reply:
@scannillo my java/kotlin mind is bugging me, but what about something like this:
```swift
class AnalyticsService {

    
    private static let instance = AnalyticsService()
    // MARK: - Internal Properties

    private var http: HTTP?
    let sessionID = UUID().uuidString.replacingOccurrences(of: "-", with: "")

    // MARK: - Initializer

    private init() {
    }
    
    public static func getInstance(http: HTTP) -> AnalyticsService {
        instance.http = http
        return instance
    }
    
    // MARK: - Internal Methods

    func sendEvent(_ name: String) async {
        let eventData = AnalyticsEventData(eventName: name, sessionID: sessionID)

        do {
            let analyticsEventRequest = try AnalyticsEventRequest(eventData: eventData)
            let (_) = try await http?.performRequest(endpoint: analyticsEventRequest)
        } catch {
            NSLog("[PayPal SDK] Failed to send analytics: %@", error.localizedDescription)
        }
    }
}
```

Reply:
> since sessionID is now a one time set in APIClient we don't need to make APIClient to be singleton

This is a good callout. You're right - technically right now, we could solve the single sessionID problem by just using a regular class with a static property for `sessionID`. The singleton setup is preparing us for when we need to take in multiple events from different feature clients, and then combine & batch them into single requests. We could 🏈 punt on it, or do it now.

Reply:
@jcnoriega - this is a cool idea! I'll give this a spin in the PR.

Reply:
Good discussion here! From my point of view, `AnalyticsService` isn't a "true" singleton and isn't the standard method for writing a singleton in Swift. @jcnoriega your code is a great option to refactor it into a "standard" singleton that manages itself. But I think the issue is that the current setup only has _features_ a singleton: there is only one, but it does not maintain its own shared instance. This was done so we could mock out `AnalyticsService` in tests.

The existing solution isn't the greatest in the world, so a discussion on improving it is great. But I think the "standard" singleton solution that Jose suggested recreates the problems of the current solution but makes it a little less readable (see `init` in the code example below). The `getInstance(http:)` method in Jose's code suggestion is, in my opinion, not very Swifty, and obfuscates how `_analyticsService` is initialized and maintained. The existing code is confusing, and in my opinion the lesser evil is to keep all of the confusing logic in one file vs having some of the logic exist in `AnalyticsService.getInstance(http:)` and some in `APIClient.init(coreConfig:)`.

We originally set this up for dependency injection into `APIClient` for testing. If we were to call `AnalyticsService.instance.sendEvent("eventName")` in every file, that would work well for running the app, but it would couple our tests to real networking logic. Even with a standard singleton solution we would still need the `weak var` with the `get` and `set` in order to mock out analytics.

All that said, with a standard singleton solution I think we would wind up with code that looks very similar to what we have now: 

```
    weak var analyticsService: AnalyticsService? {
        get { APIClient._analyticsService }
        set { APIClient._analyticsService = newValue }
    }
    private static var _analyticsService: AnalyticsService?
    
    // MARK: - Public Initializer

    public init(coreConfig: CoreConfig) {
        self.http = HTTP(coreConfig: coreConfig)
        self.coreConfig = coreConfig
        APIClient._analyticsService = AnalyticsService.getInstance(http)
    }
```

Reply:
Also to clarify, my plan was that the feature clients won't interact with `AnalyticsService` at all, but instead just call `APIClient.sendEvent` ([code](https://github.com/paypal/iOS-SDK/blob/8f8354122809ce9c2eea9e8552b14e48c3fc20a3/Sources/PaymentsCore/Networking/APIClient.swift#L55)).

Also in both the current and Jose's suggested implementation, we'll be able to inject a mock `HTTP` into `AnalyticsService` so we can mock the actual network request happening.

Reply:
> The getInstance(http:) method in Jose's code suggestion is, in my opinion, not very Swifty, and obfuscates how _analyticsService is initialized and maintained.

I agree this method/style isn't Swifty. But there's also just been a lot of general confusion regarding the `_propertyName` syntax with an internal private var like the current implementation. I think Jose's suggestion ends up providing the same solution and is a little easier to wrap your head around.